warehouse/accounts/views.py

```diff
@@ -122,38 +123,102 @@ def login(request, redirect_field_name=REDIRECT_FIELD_NAME, _form_class=LoginFor
             username = form.username.data
             userid = user_service.find_userid(username)
 
+            # If the user has enabled two factor authentication.
+            if user_service.has_two_factor(userid):
+                two_factor_data = {"userid": userid}
+                if redirect_to:
+                    two_factor_data["redirect_to"] = redirect_to
+
+                # Save token to cookies and redirect to two-factor page.
+                resp = HTTPSeeOther(
+                    request.route_path("accounts.two-factor", _query=two_factor_data)
+                )
+
+                return resp
+            else:
+                # If the user-originating redirection url is not safe, then
+                # redirect to the index instead.
+                if not redirect_to or not is_safe_url(
+                    url=redirect_to, host=request.host
+                ):
+                    redirect_to = request.route_path("manage.projects")
+
+                # Actually perform the login routine for our user.
+                headers = _login_user(request, user_service, userid)
+
+                # Now that we're logged in we'll want to redirect the user to
+                # either where they were trying to go originally, or to the default
+                # view.
+                resp = HTTPSeeOther(redirect_to, headers=dict(headers))
+
+                # We'll use this cookie so that client side javascript can
+                # Determine the actual user ID (not username, user ID). This is
+                # *not* a security sensitive context and it *MUST* not be used
+                # where security matters.
+                #
+                # We'll also hash this value just to avoid leaking the actual User
+                # IDs here, even though it really shouldn't matter.
+                resp.set_cookie(
+                    USER_ID_INSECURE_COOKIE,
+                    hashlib.blake2b(str(userid).encode(), person=b"warehouse.userid")
+                    .hexdigest()
+                    .lower(),
+                )
+
+            return resp
+
+    return {
+        "form": form,
+        "redirect": {"field": REDIRECT_FIELD_NAME, "data": redirect_to},
+    }
+
+
+@view_config(
+    route_name="accounts.two-factor",
+    renderer="accounts/two-factor.html",
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+)
+def two_factor(request, _form_class=TwoFactorForm):
+    if request.authenticated_userid is not None:
+        return HTTPSeeOther(request.route_path("manage.projects"))
+
+    userid = request.params.get("userid")
```

Comment:
I *think* we need to sign this data (or stash it in the server side session data, but signing is probably good enough) because right now I think this allows people to degrade 2fa auth into a single factor (the TOTP token). Basically this currently lets someone just hit this URL directly and completely bypass the need to type in a password at all (unless I'm misinterpreting this?).

Perhaps It would be better to roll this page into the normal login page, and make the login page do something like:

1. If Get, render the 1fa log in form.
2. If Post, validate the 1fa login form.
  1. If Invalid, render 1fa login form with error.
  2. If valid, check if user has 2fa, if so render 2fa login form (subclass of 1fa login form) with username/pw already filled out.

Does that make sense? That removes the need to shuffle data around at all, and means that we always check if the username, password and totp values are all valid.

One possible riff on this idea is to check if the user has 2fa enabled before we check the password at all, and if they do render the 2fa form first, and skip validating the pw in that case. That means we would leak the information if a person has TOTP enabled or not, but it also means you can't brute force password validation I think? 

Thoughts?

Reply:
> Basically this currently lets someone just hit this URL directly and completely bypass the need to type in a password at all (unless I'm misinterpreting this?).

Yeah, this looks correct. Major oversight on my part -- I thought that `uses_session` on a `view_config` provided implicit integrity, but that's clearly wrong 🤐

Any pointers on signing query parameters (is this something we can do?), or are you thinking we should just revert to the earlier signed cookie version?

CCing @nlhkabu for thoughts on merging this with the normal login page: my big concern there is that things will get complicated quickly as we add multiple second factor types.

Re: Checking 2FA first: I can't think of any security reason (other than the information leakage) why this wouldn't be fine, although I've never seen it done before. Password bruteforcing should already be capped by your rate limiter, so I would err on the conservative side here and suggest that we only test the second factor after confirming the user's password.

Reply:
Ah yea, ``uses_session`` just controls whether you can access ``request.session`` or not. Anything stored inside of ``request.session`` is securely tied to that specific session, but beyond that, ``uses_session`` doesn't have any real security implications. It exists primarily for caching (anything that uses the session can't be cached globally).

We've never put any signed content into the URL before, but you can use ``warehouse.utils.crypto`` to get a signer that will let you sign an opaque string.

The other option is to just stick the user id in ``request.session`` in the login view before we redirect to the 2fa page. That allows us not to have any need for crypto in the login flow, but it does mean that our login flow becomes stateful server side which is maybe kind of weird? I'm not sure which option I prefer, but I think I prefer the crypto solution.

Reply:
> One possible riff on this idea is to check if the user has 2fa enabled before we check the password at all, and if they do render the 2fa form first, and skip validating the pw in that case. That means we would leak the information if a person has TOTP enabled or not, but it also means you can't brute force password validation I think?

If we don't ask for a password, would this not just be a single factor login?  

> thoughts on merging this with the normal login page: my big concern there is that things will get complicated quickly as we add multiple second factor types

My strong preference is to keep these pages separate, as this is the typical work flow on other 2fa interfaces.